crates/hyperdrive-math/src/lp.rs

```diff
@@ -1,12 +1,161 @@
+use eyre::{eyre, Result};
 use std::cmp::Ordering;
 
 use ethers::types::{I256, U256};
 use fixed_point::FixedPoint;
 use fixed_point_macros::{fixed, int256};
 
-use crate::{State, YieldSpace};
+use crate::{calculate_effective_share_reserves, State, YieldSpace};
 
 impl State {
+    // Calculates the lp_shares for a given contribution when adding liquidity.
+    pub fn calculate_add_liquidity(
+        &self,
+        current_block_timestamp: U256,
+        contribution: FixedPoint,
+        min_lp_share_price: FixedPoint,
+        min_apr: FixedPoint,
+        max_apr: FixedPoint,
+        as_base: bool,
+    ) -> Result<FixedPoint> {
+        // Ensure that the contribution is greater than or equal to the minimum
+        // transaction amount.
+        if contribution < self.minimum_transaction_amount() {
+            return Err(eyre!(
+                "MinimumTransactionAmount: Contribution is smaller than the minimum transaction."
+            ));
+        }
+
+        // Enforce the slippage guard.
+        let apr = self.calculate_spot_rate();
+        if apr < min_apr || apr > max_apr {
+            return Err(eyre!("InvalidApr: Apr is outside the slippage guard."));
+        }
+
+        // Get lp_total_supply for the lp_shares calculation.
+        let lp_total_supply = self.lp_total_supply();
```

Comment:
No, I thought so too, but if you look in HyperdriveTarget0.sol, we return this formula as lpTotalSupply in PoolInfo.  Since `self.lp_total_supply()` is reading from pool info, we assume this formula is baked in.